gen/map.go

```diff
@@ -281,3 +281,144 @@ func (m *mapGenerator) equalsUnhashable(g Generator, spec *compile.MapSpec) (str
 
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
+
+func (m *mapGenerator) zapMarshaler(
+	g Generator,
+	spec compile.TypeSpec,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	name := zapperName(g, spec)
+	switch root.KeySpec.(type) {
+	case *compile.StringSpec:
+		// We are already at the root, so no need to check if it is a typedef of a
+		// string. For simplicity, we always cast the key to a string when logging this way.
+		if err := g.EnsureDeclared(
+			`
+				type <.Name> <typeReference .Type>
+				<$zapcore := import "go.uber.org/zap/zapcore">
+				<$m := newVar "m">
+				<$k := newVar "k">
+				<$v := newVar "v">
+				<$enc := newVar "enc">
+				// MarshalLogObject implements zapcore.ObjectMarshaler, enabling
+				// fast logging of <.Name>.
+				func (<$m> <.Name>) MarshalLogObject(<$enc> <$zapcore>.ObjectEncoder) error {
+					for <$k>, <$v> := range <$m> {
+						<- $encAdd := printf "%s.Add%s((string)(%s), %s)" $enc (zapEncoder .Type.ValueSpec) $k (zapMarshaler .Type.ValueSpec $v)>
+						<if (zapCanError .Type.ValueSpec) ->
+							if err := <$encAdd>; err != nil {
+								return err
```

Comment:
Damn, this is getting repetitive. I wonder if we can clean up this `err != nil` business.

Can we maybe do something like,

```
<zapEncodeBegin .Type.ValueSpec>
  <$enc>.Add<zapEncoder ..>(..., <zapMarshaler .Type.ValueSpec $v>)
<zapEncodeEnd .Type.ValueSpec>
```

Where zapEncodeBegin and zapEncodeEnd add the `if err != nil`, etc. if needed?

This will handle all these cases where we're doing `if zapCanErr` and
`$errAdd` again and again.